src/lib/io/sol-uart-impl-linux.c

```diff
@@ -172,72 +293,85 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
 SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
+    struct sol_blob *blob;
+    uint16_t i;
+
     SOL_NULL_CHECK(uart);
-    if (uart->async.rx_fd_handler)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
+
+    if (uart->in_use) {
+        uart->delete_me = true;
+        return;
+    }
+
+    uart->in_use = true;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        if (uart->tx_cb)
+            uart->tx_cb((void *)uart->user_data, uart, blob, -ECANCELED);
+        sol_blob_unref(blob);
+    }
+
+    if (uart->rx.used)
+        uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+
+    if (uart->fd_handler)
+        sol_fd_del(uart->fd_handler);
+
+    if (uart->read_timeout)
+        sol_timeout_del(uart->read_timeout);
```

Comment:
maybe instead of checking `delete_me` in the fd handler and timer as I suggest, just move these up and always delete them. Just remember to set pointers to NULL. What must be protected is the `pending_blobs` iteration and the `uart` handle itself.